From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 16 Aug 2025 20:29:29 -0700
Subject: [PATCH] Filter ClientboundSetEntityMotionPacket

Original license: GPL v3
Original project: https://github.com/CraftCanvasMC/Canvas

Entities with a 360-degree rotation and movement, and item and ender eye entities
get royally screwed over with this change, as it removes the synchronization
of the entity delta movement.

This packet, when disabled, can reduce bandwidth by 60%, which is insane, but we do
need to add a rough filter that ensures there is no visual impact on the client

The only entities needing this are:
1. Squid, only mobs with 360° range of rotation
2. Item Entities, disabling this causes glitches and often slightly desyncs their position within 0.5 blocks on the client side. But we filter it if
   it hasn't moved, since for some reason it sends those packets anyway???
3. Ender Eyes, they depend on their velocity being set bc the packet sets the velocity on the client, but the client doesn't handle the
     ender eye velocity on its own bc the ender eye velocity is in relation to the stronghold structure which the client doesn't have access to
4. Shulker bullets use random to determine where they are going, so we need to send velocity packets to ensure it remains smooth on the client

This filter helps keep unneeded packets reduced, but also ensures that we keep smooth visual gameplay

diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 057e38b58454a53af3d4bb26dfe95a10231c9456..8911047b4b50a808c71596e6a3daa53a12417c60 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -221,7 +221,12 @@ public class ServerEntity {
                                         )
                                     )
                                 );
-                        } else { // Leaf - Multithreaded tracker - diff on change
+                        } else if (!org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.filterClientboundSetEntityMotionPacket || // Canvas - filter ClientboundSetEntityMotionPacket
+                            (this.entity instanceof net.minecraft.world.entity.item.ItemEntity && flag1) ||
+                            this.entity instanceof net.minecraft.world.entity.projectile.EyeOfEnder ||
+                            this.entity instanceof net.minecraft.world.entity.animal.squid.Squid ||
+                            this.entity instanceof net.minecraft.world.entity.projectile.ShulkerBullet
+                        ) { // Leaf - Multithreaded tracker - diff on change
                             this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
                         }
                     }
@@ -555,7 +560,12 @@ public class ServerEntity {
                                     new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement),
                                     new ClientboundProjectilePowerPacket(abstractHurtingProjectile.getId(), abstractHurtingProjectile.accelerationPower)
                                 )));
-                            } else {
+                            } else if (!org.dreeam.leaf.config.modules.opt.ReduceUselessPackets.filterClientboundSetEntityMotionPacket || // Canvas - filter ClientboundSetEntityMotionPacket
+                                (this.entity instanceof net.minecraft.world.entity.item.ItemEntity && flag1) ||
+                                this.entity instanceof net.minecraft.world.entity.projectile.EyeOfEnder ||
+                                this.entity instanceof net.minecraft.world.entity.animal.squid.Squid ||
+                                this.entity instanceof net.minecraft.world.entity.projectile.ShulkerBullet
+                            ) {
                                 ctx.sendToTrackingPlayers(tracker, new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
                             }
                         }
